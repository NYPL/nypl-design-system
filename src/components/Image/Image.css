.image {
  display: block;
  height: auto;
  max-width: 100%;

  &--circle {
    border-radius: 50%;
  }
}

.nypl-ds {
  figure {
    margin: unset;
  }
}

.figure {
  &__image {
    margin-bottom: var(--space-xxs);

    &--circle {
      border-radius: 50%;
    }
  }

  &__figcaption {
    font-style: italic;
  }
}

.figcaption {
  &__credit,
  &__caption {
    font-size: var(--font-size--1);
    margin-bottom: var(--space-xxs);
  }
}
