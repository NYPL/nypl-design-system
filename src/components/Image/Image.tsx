// Images
import * as React from "react";
import bem from "../../utils/bem";

export interface ImageProps {
  /** Text description of the image */
  alt: string;
  /** Additional attributes passed to the image */
  attributes?: { [key: string]: any };

  /** BlockName for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** ClassName you can add in addition to 'image' */
  className?: string;
  /** Adding will wrap the image in a <figure> */
  imageCaption?: string;
  /** Adding will wrap the image in a <figure> */
  imageCredit?: string;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** The src attribute is required, and contains the path to the image you want to embed. */
  src: string;
}

export default function Image(props: React.ComponentProps<"img"> & ImageProps) {
  const baseClass = "image";
  const {
    alt,
    attributes,
    blockName,
    className,
    imageCaption,
    imageCredit,
    modifiers = [],
    src,
    ...imgHTMLProps
  } = props;

  if (alt && alt.length > 300) {
    throw new Error("Alt Text must be less than 300 characters");
  }

  const figureBlockName = imageCaption || imageCredit ? "figure" : blockName;

  const imageProps: ImageProps = {
    className: bem(baseClass, modifiers, figureBlockName, [className]),
    src: src,
    alt: alt,
  };

  return (
    <>
      {imageCaption || imageCredit ? (
        <figure className={bem("figure")}>
          <img
            alt={imageProps.alt}
            {...imageProps}
            {...imgHTMLProps}
            {...attributes}
          />
          <figcaption className={bem("figcaption", [], "figure", ["image"])}>
            {imageCaption && (
              <div className={bem("caption", [], "figcaption")}>
                {imageCaption}
              </div>
            )}
            {imageCredit && (
              <div className={bem("credit", [], "figcaption")}>
                {imageCredit}
              </div>
            )}
          </figcaption>
        </figure>
      ) : (
        <img
          alt={imageProps.alt}
          {...imageProps}
          {...imgHTMLProps}
          {...attributes}
        />
      )}
    </>
  );
}
