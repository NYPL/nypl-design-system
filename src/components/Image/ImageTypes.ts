export enum ImageRatios {
  FourByThree = "fourByThree",
  OneByTwo = "oneByTwo",
  Original = "original",
  SixteenByNine = "sixteenByNine",
  Square = "square",
  ThreeByFour = "threeByFour",
  ThreeByTwo = "threeByTwo",
  TwoByOne = "twoByOne",
}

export enum ImageSizes {
  Default = "default",
  Large = "large",
  Medium = "medium",
  Small = "small",
  ExtraSmall = "xsmall",
  ExtraExtraSmall = "xxsmall",
}

export enum ImageTypes {
  Default = "default",
  Circle = "circle",
}
