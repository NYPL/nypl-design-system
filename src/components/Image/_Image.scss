.image {
  display: block;
  height: auto;
  max-width: 100%;

  &--circle {
    border-radius: 50%;
  }
}

.nypl-ds {
  figure {
    margin: unset;
  }
}

.figure {
  &__image {
    @include space-stack-xxs;

    &--circle {
      border-radius: 50%;
    }
  }

  &__figcaption {
    font-style: italic;
  }
}

.figcaption {
  &__credit,
  &__caption {
    @include space-stack-xxs;

    font-size: var(--font-size--1);
  }
}
