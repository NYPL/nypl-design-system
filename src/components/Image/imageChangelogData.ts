/** This data is used to populate the ComponentChangelogTable component.
 *
 * date: string (when adding new entry during development, set value as "Prerelease")
 * version: string (when adding new entry during development, set value as "Prerelease")
 * type: "Bug Fix" | "New Feature" | "Update";
 * affects: array["Accessibility" | "Documentation" | "Functionality" | "Styles"];
 * notes: array (will render as a bulleted list, add one array element for each list element)
 */
import { ChangelogData } from "../../utils/ComponentChangelogTable";

export const changelogData: ChangelogData[] = [
  {
    date: "Prerelease",
    version: "Prerelease",
    type: "Bug Fix",
    affects: ["Accessibility"],
    notes: [
      "Fixes an issue where the `id` prop was being duplicated in the aspect ratio's wrapper div.",
    ],
  },
  {
    date: "2024-08-29",
    version: "3.3.0",
    type: "Update",
    affects: ["Accessibility"],
    notes: [
      "Defaults the `alt` attribute to an empty string if no value is passed.",
    ],
  },
  {
    date: "2024-04-25",
    version: "3.1.1",
    type: "Update",
    affects: ["Functionality"],
    notes: ["Adds `fallbackSrc` and `onError` props."],
  },
  {
    date: "2024-03-14",
    version: "3.0.0",
    type: "Update",
    affects: ["Styles"],
    notes: ["Chakra 2.8 update."],
  },
  {
    date: "2023-12-07",
    version: "2.1.3",
    type: "Update",
    affects: ["Documentation", "Functionality", "Styles"],
    notes: ["Adds `sizeBasedOn` prop to the `Logo` component."],
  },
];
