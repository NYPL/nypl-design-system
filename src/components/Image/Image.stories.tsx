import * as React from "react";

import Image from "./Image";
import { text, boolean, select } from "@storybook/addon-knobs";

export default {
  title: "Image",
  component: Image,
};

let showCaption, showCredit;

const imageRatios = {
  "1x1": "https://placeimg.com/100/100/animals",
  "2x1": "https://placeimg.com/200/100/animals",
  "3x4": "https://placeimg.com/150/200/animals",
  "4x1": "https://placeimg.com/400/100/animals",
  "4x3": "https://placeimg.com/200/150/animals",
  "16x9": "https://placeimg.com/400/225/animals",
};

export const ImageWithOptionalCaptionandCredit = () => (
  <>
    {boolean("Show Caption", true)
      ? (showCaption = true)
      : (showCaption = false)}
    {boolean("Show Credit", true) ? (showCredit = true) : (showCredit = false)}
    <Image
      src={select("Initial Selected Option", imageRatios, imageRatios["2x1"])}
      imageCaption={
        showCaption
          ? text("Image Caption", "Deserted Islander Relocator")
          : null
      }
      imageCredit={
        showCredit ? text("Image Credit", "Taken by Tom Nook") : null
      }
      modifiers={boolean("Apply Circle Mask", false) ? ["circle"] : null}
      alt={text("Image alt", "alt")}
    />
  </>
);

ImageWithOptionalCaptionandCredit.storyName =
  "Image with Optional Caption and Credit";
ImageWithOptionalCaptionandCredit.parameters = {
  design: {
    type: "figma",
    url:
      "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=11896%3A45379",
  },
};
