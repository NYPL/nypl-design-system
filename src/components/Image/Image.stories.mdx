import { VStack } from "@chakra-ui/react";
import {
  ArgsTable,
  Canvas,
  Description,
  Meta,
  Story,
} from "@storybook/addon-docs";
import { withDesign } from "storybook-addon-designs";

import Heading from "../Heading/Heading";
import Image from "./Image";
import { ImageRatios, ImageSizes, ImageTypes } from "./ImageTypes";
import { getCategory } from "../../utils/componentCategories";
import { getStorybookEnumValues } from "../../utils/utils";

export const ratiosEnumValues = getStorybookEnumValues(
  ImageRatios,
  "ImageRatios"
);
export const sizesEnumValues = getStorybookEnumValues(ImageSizes, "ImageSizes");
export const typesEnumValues = getStorybookEnumValues(ImageTypes, "ImageTypes");

<Meta
  title={getCategory("Image")}
  component={Image}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=11896%3A45379",
    },
    jest: ["Image.test.tsx"],
  }}
  argTypes={{
    additionalStyles: { control: false },
    alt: {
      control: false,
      description:
        "Alternate text description of the image. Needed for accessibility purposes.",
    },
    className: {
      control: false,
      description: "Additional class name for the component to use.",
    },
    component: { control: false },
    src: {
      description:
        "The src attribute is required, and contains the path to the image you want to embed.",
    },
  }}
/>

# Image

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.0.6`    |
| Latest            | `0.25.3`   |

<Description of={Image} />

<p>
  If you want a simple HTML `img` element then don't pass in values for the
  `imageAspectRatio` or `imageSize` props.
</p>

<Canvas>
  <Story
    name="Image with Controls"
    args={{
      alt: "Alt text",
      imageAspectRatio: "ImageRatios.TwoByOne",
      imageCaption: "Image caption",
      imageCredit: "Image credit",
      imageSize: "ImageSizes.Medium",
      imageType: "ImageTypes.Default",
      src: "https://placeimg.com/400/300/animals",
    }}
    argTypes={{
      imageAspectRatio: {
        control: { type: "select" },
        table: { defaultValue: { summary: "ImageRatios.Original" } },
        options: ratiosEnumValues.options,
      },
      imageSize: {
        control: { type: "select" },
        table: { defaultValue: { summary: "ImageSizes.Default" } },
        options: sizesEnumValues.options,
      },
      imageType: {
        control: { type: "select" },
        table: { defaultValue: { summary: "ImageTypes.Default" } },
        options: typesEnumValues.options,
      },
    }}
  >
    {(args) => (
      <Image
        {...args}
        imageAspectRatio={ratiosEnumValues.getValue(args.imageAspectRatio)}
        imageSize={sizesEnumValues.getValue(args.imageSize)}
        imageType={typesEnumValues.getValue(args.imageType)}
      />
    )}
  </Story>
</Canvas>

<ArgsTable story="Image with Controls" />

## Image Figure

By passing in `imageCredit` or `imageCaption`, the `Image` component will
render an HTML `figure` element and render an `img` element inside with the
proper props. If no `imageCredit` or `imageCaption` prop values are passed,
then an `img` element will be rendered with or without wrapper style divs.

<Canvas>
  <Story
    name="Figure and figcaption"
    args={{
      alt: "Alt text",
      imageAspectRatio: ImageRatios.ThreeByTwo,
      imageCaption: "Image caption",
      imageCredit: "Image credit",
      imageSize: ImageSizes.Medium,
      imageType: ImageTypes.Default,
      src: "https://placeimg.com/400/300/animals",
    }}
    argTypes={{
      imageAspectRatio: { table: { disable: true } },
      imageCaption: { table: { disable: true } },
      imageCredit: { table: { disable: true } },
      imageSize: { table: { disable: true } },
      imageType: { table: { disable: true } },
      src: { table: { disable: true } },
    }}
  >
    {(args) => <Image {...args} />}
  </Story>
</Canvas>

## Image Sizes

Use the `ImageSizes` prop to set the desired size.

<Canvas>
  <Story name="Image Sizes">
    <VStack spacing="xs">
      <Heading id="Small" text="ImageSizes.Small" />
      <Image
        alt="Alt text"
        imageSize={ImageSizes.Small}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="Medium" text="ImageSizes.Medium" />
      <Image
        alt="Alt text"
        imageSize={ImageSizes.Medium}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="Large" text="ImageSizes.Large" />
      <Image
        alt="Alt text"
        imageSize={ImageSizes.Large}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="Default" text="ImageSizes.Default" />
      <Image alt="Alt text" src="https://placeimg.com/400/300/animals" />
    </VStack>
  </Story>
</Canvas>

## Image Aspect Ratios

Use the `imageAspectRatio` prop to set the desired aspect ratio. Note: the
following example has `imageSize` set to `ImageSizes.Small`.

For a better viewing experience, the `Image` components below have been wrapped
in an element with a fixed width value.

<Canvas>
  <Story name="Image Aspect Ratios">
    <VStack spacing="xs" maxWidth="360px" margin="auto">
      <Heading id="fourbythree" text="ImageRatios.FourByThree" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.FourByThree}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="onebytwo" text="ImageRatios.OneByTwo" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.OneByTwo}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="original" text="ImageRatios.Original" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.Original}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="sixteenbynine" text="ImageRatios.SixteenByNine" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.SixteenByNine}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="square" text="ImageRatios.Square" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.Square}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="threebyfour" text="ImageRatios.ThreeByFour" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.ThreeByFour}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="threebytwo" text="ImageRatios.ThreeByTwo" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.ThreeByTwo}
        src="https://placeimg.com/400/300/animals"
      />
      <Heading id="twobyone" text="ImageRatios.TwoByOne" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.TwoByOne}
        src="https://placeimg.com/400/300/animals"
      />
    </VStack>
  </Story>
</Canvas>

## Image Types

Use the `imageType` prop to get an image variant. Note that to get a perfect
circle, the `imageAspectRatio` prop _must_ be set to `ImageRatios.Square`.

For a better viewing experience, the `Image` components below have been wrapped
in an element with a fixed width value.

<Canvas>
  <Story name="Image Types">
    <VStack spacing="xs" maxWidth="360px" margin="auto">
      <Heading id="default" text="ImageTypes.Default" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.Square}
        imageType={ImageTypes.Default}
        src="https://placeimg.com/400/400/animals"
      />
      <Heading id="circle" text="ImageTypes.Circle" />
      <Image
        alt="Alt text"
        imageAspectRatio={ImageRatios.Square}
        imageType={ImageTypes.Circle}
        src="https://placeimg.com/400/400/animals"
      />
    </VStack>
  </Story>
</Canvas>
