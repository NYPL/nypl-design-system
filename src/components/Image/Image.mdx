import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import ComponentChangelogTable from "../../utils/ComponentChangelogTable";
import * as ImageStories from "./Image.stories";
import Link from "../Link/Link";
import { changelogData } from "./imageChangelogData";

<Meta of={ImageStories} />

# Image

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.0.6`    |
| Latest            | `3.3.1`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#credit-and-caption" target="_self">Credit and Caption</Link>}
- {<Link href="#sizes" target="_self">Sizes</Link>}
- {<Link href="#aspect-ratios" target="_self">Aspect Ratios</Link>}
- {<Link href="#types" target="_self">Types</Link>}
- {<Link href="#html-attributes" target="_self">HTML Attributes</Link>}
- {<Link href="#fallback-image" target="_self">Fallback Image</Link>}
- {<Link href="#lazy-loading" target="_self">Lazy Loading</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

<Description of={ImageStories.WithControls} />

If you want a simple HTML `img` element then don't pass in values for the
`aspectRatio` or `size` props.

## Component Props

<Canvas of={ImageStories.WithControls} />

<ArgTypes of={ImageStories.WithControls} />

## Accessibility

The `alt` prop should be used to concisely describe the image. If the image is purely decorative,
the `alt` attribute should still be present, but can be set to an empty string. In this scenario,
an `alt` prop does not need to be passed. The component will set the attribute to an empty string
`""` by default.

Resources:

- [W3C WAI Images Tutorial](https://www.w3.org/WAI/tutorials/images/)
- [WebAIM Accessible Images](https://webaim.org/techniques/images/)
- [Yale Usability & Web Accessibility](https://usability.yale.edu/web-accessibility/articles/images)

## Credit and Caption

By passing in `credit` or `caption`, the `Image` component will
render an HTML `figure` element and render an `img` element inside with the
proper props. If no `credit` or `caption` prop values are passed,
then an `img` element will be rendered with or without wrapper style divs.

<Canvas of={ImageStories.FigureAndFigcaption} />

## Sizes

Use the `size` prop to set the desired size with the `ImageSizes` enum.

In addition, the `sizeBasedOn` prop can be passed to determine whether the size
of the `Image` is updated based on either the `"height"` or `"width"`. By default,
the `sizeBasedOn` prop is set to `"width"`.

### Size Based On Width

<Canvas of={ImageStories.Sizes} />

### Size Based On Height

<Canvas of={ImageStories.SizesBasedOnHeight} />

## Aspect Ratios

Use the `aspectRatio` prop to set the desired aspect ratio. Note: the
following example has `size` set to `ImageSizes.Small`.

For a better viewing experience, the `Image` components below have been wrapped
in an element with a fixed width value.

<Canvas of={ImageStories.AspectRatios} />

## Types

Use the `imageType` prop to get an image variant. Note that to get a perfect
circle, the `aspectRatio` prop _must_ be set to `"square"`.

For a better viewing experience, the `Image` components below have been wrapped
in an element with a fixed width value.

<Canvas of={ImageStories.Types} />

## HTML Attributes

The `Image` component accepts HTML image attributes. If needed, you can pass
functions such as `onLoad` and `onError` or attributes such as `srcset`.

For both examples below, open the browser's console to see the log messages. In
the first example below, when the image is finished loading, the `onLoad`
function callback is called. The image's dimensions are logged. In the second
example, there is an (intentional) error with the image's URL. This will trigger
the `onError` function callback.

<Canvas of={ImageStories.HTMLAttributes} />

## Fallback Image

Sometimes images don't load due to network issues or the image URL is incorrect.
Set a fallback image source through the `fallbackSrc` prop. Internally, the
`Image` component will call the `img` element's `onerror` attribute with a
callback function to load the `fallbackSrc` image. Additionally, a warning will
be logged to the console when the error occurs and the fallback image is loaded.

If an _additional_ action needs to be performed when the fallback image is
loaded, pass a callback function to the `onError` prop.

<Source
  code={`
<Image
  alt="Alt text"
  fallbackSrc="https://loremflickr.com/540/420/cat"
  onError={(event) => console.log("Fallback image loaded"); }
  src="https://placekitten.com/540/420"
/>
  `}
  language="jsx"
/>

**IMPORTANT**: If a custom image component, such as the Next.js `Image` component,
is passed to the `component` prop, then the `fallbackSrc` and `onError` props
won't work. It is expected that the custom image component that is passed in
implements the `onError` attribute.

<Source
  code={`
import { Image } from "@nypl/design-system-react-components";
import Image as NextImage from "next/image";

// ...

<Image
  component={{
    <NextImage
      src="https://placekitten.com/540/420"
      onError={(event) => {
        // Implement your own fallback event
        event.target.src =  "https://loremflickr.com/540/420/cat"
      }}
    />
  }} 
/>
`}
  language="jsx"
/>

<Canvas of={ImageStories.FallbackImage} />

Resources:

- [MDN Image loading errors](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#image_loading_errors)
- [MDN HTMLElement: error event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/error_event)

## Lazy Loading

The `Image` component handles lazy loading images through two npm packages:
[`react-intersection-observer`](https://github.com/thebuilder/react-intersection-observer)
and [`@charlietango/use-native-lazy-loading`](https://www.npmjs.com/package/@charlietango/use-native-lazy-loading).
The first package handles using the `IntersectionObserver` API in React and the
second package checks if the `img`'s native `loading="lazy"` attribute is
supported by the browser.

Lazy loading is disabled by default. In order to use the lazy loading feature,
pass the `isLazy` prop or set it to `true`.

<Source
  code={`
<Image alt="Alt text" isLazy src="https://loremflickr.com/540/420/new+york+public+library" />
`}
  language="jsx"
/>

The `Image`'s implementation of lazy loading is based on the recipe provided by
`react-intersection-observer`. Once the image's wrapper is in view, we load the
image by rendering it through the native `loading="lazy"` attribute or through
the `IntersectionObserver` API.

In the following example, open the "Network" tab in your browser's console and
as you scroll down to view the next set of 10 images, you'll see that they get
loaded as the images come into the window's view.

Resources:

- [Browser-level image lazy-loading for the web](https://web.dev/browser-level-image-lazy-loading/)

<Canvas of={ImageStories.LazyLoading} />

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
