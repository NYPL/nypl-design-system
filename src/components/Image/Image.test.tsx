import { expect } from "chai";
import * as Enzyme from "enzyme";
import * as React from "react";

import Image from "./Image";

describe("Images", () => {
  let wrapper: Enzyme.ShallowWrapper<any, any>;
  const tooManyChars =
    "UNwbNIp1GFehVO1LBBIFYF6Nv5IhHeCPgW5JHet2VRuuKnKlGxSgjbFcfCbcttVTkxbE7ItibBIuG9cTA9HGntZSKv31u7egV88bIg7DGDHmi2nKq3ssZkBDl7oWfYpyjWbfV4kZx9EDPsKHUTKb2xIKZWSbRMuTUtsJWonQyfitDp8ui1P7QR80LA236yE5fCkRImbltEsqJWEYxyotQpLjFbruJsPcGSCp6ET6DCrNQeWFsRVaM2Co99ewZjLuY42kdpBEXjcw9HPcTjKKZw141sKBNOoFfNMueYaHtNjNI";
  it("Shows Image", () => {
    wrapper = Enzyme.shallow(<Image src="test.png" alt={""} />);
    expect(wrapper.find("img")).to.have.lengthOf(1);
  });
  it("Shows Image wrapped in figure when provided ImageCaption", () => {
    wrapper = Enzyme.shallow(
      <Image src="test.png" imageCaption={"caption"} alt={""} />
    );
    expect(wrapper.find("figure")).to.have.lengthOf(1);
    expect(wrapper.find("figcaption")).to.have.lengthOf(1);
  });

  it("Shows Image wrapped in figure when provided ImageCredit", () => {
    wrapper = Enzyme.shallow(
      <Image src="test.png" imageCredit={"credit"} alt={""} />
    );

    expect(wrapper.find("figure")).to.have.lengthOf(1);
    expect(wrapper.find("figcaption")).to.have.lengthOf(1);
  });

  it("Shows Image wrapped in figure with credit and caption when provided ImageCredit and ImageCaption", () => {
    wrapper = Enzyme.shallow(
      <Image
        src="test.png"
        imageCaption={"caption"}
        imageCredit={"credit"}
        alt={""}
      />
    );
    expect(wrapper.find("figcaption")).to.have.lengthOf(1);
    expect(wrapper.find("div")).to.have.lengthOf(2);
  });

  it("Throws error when alt text is too long", () => {
    expect(() =>
      Enzyme.mount(<Image src="test.png" alt={tooManyChars} />)
    ).to.throw("Alt Text must be less than 300 characters");
  });
});
