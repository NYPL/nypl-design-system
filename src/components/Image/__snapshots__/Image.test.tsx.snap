// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Renders the UI snapshot correctly 1`] = `
<img
  alt=""
  className="css-0"
  src="test.png"
/>
`;

exports[`Image Renders the UI snapshot correctly 2`] = `
<figure
  className="css-0"
>
  <img
    alt=""
    className="css-0"
    src="test.png"
  />
  <figcaption
    className="css-0"
  >
    <div
      className="css-0"
    >
      Caption
    </div>
  </figcaption>
</figure>
`;

exports[`Image Renders the UI snapshot correctly 3`] = `
<img
  alt=""
  className="css-0"
  src="test.png"
/>
`;

exports[`Image Renders the UI snapshot correctly 4`] = `
<img
  alt=""
  className="css-0"
  src="test.png"
/>
`;

exports[`Image Renders the UI snapshot correctly 5`] = `
<img
  alt=""
  className="css-0"
  src="test.png"
/>
`;

exports[`Image Renders the UI snapshot correctly 6`] = `
<img
  alt=""
  className="css-0"
  src="test.png"
/>
`;

exports[`Image Renders the UI snapshot correctly 7`] = `
<img
  alt=""
  className="css-0"
  src="test.png"
/>
`;

exports[`Image Renders the UI snapshot correctly 8`] = `
<div
  className="the-wrap  css-0"
>
  <div
    className="the-crop css-0"
  >
    <img
      alt=""
      className="css-0"
      src="test.png"
    />
  </div>
</div>
`;

exports[`Image Renders the UI snapshot correctly 9`] = `
<div
  className="the-wrap  css-0"
>
  <div
    className="the-crop css-0"
  >
    <img
      alt=""
      className="css-0"
      src="test.png"
    />
  </div>
</div>
`;

exports[`Image Renders the UI snapshot correctly 10`] = `
<img
  alt=""
  className="css-0"
  src="test.png"
/>
`;

exports[`Image Renders the UI snapshot correctly 11`] = `
<div
  className="the-wrap  css-0"
>
  <div
    className="the-crop css-0"
  >
    <img
      alt=""
      className="css-0"
      src="test.png"
    />
  </div>
</div>
`;

exports[`Image Renders the UI snapshot correctly 12`] = `
<div
  className="the-wrap  css-0"
>
  <div
    className="the-crop css-0"
  >
    <img
      alt=""
      className="css-0"
      src="test.png"
    />
  </div>
</div>
`;

exports[`Image Renders the UI snapshot correctly 13`] = `
<div
  className="the-wrap  css-0"
>
  <div
    className="the-crop css-0"
  >
    <img
      alt=""
      className="css-0"
      src="test.png"
    />
  </div>
</div>
`;

exports[`Image Renders the UI snapshot correctly 14`] = `
<div
  className="the-wrap  css-0"
>
  <div
    className="the-crop css-0"
  >
    <img
      alt=""
      className="css-0"
      src="test.png"
    />
  </div>
</div>
`;

exports[`Image Renders the UI snapshot correctly 15`] = `
<div
  className="the-wrap  css-0"
>
  <div
    className="the-crop css-0"
  >
    <img
      alt=""
      className="css-0"
      src="test.png"
    />
  </div>
</div>
`;

exports[`Image Renders the UI snapshot correctly 16`] = `
<div
  className="the-wrap  css-0"
>
  <div
    className="the-crop css-0"
  >
    <img
      alt=""
      className="css-0"
      src="test.png"
    />
  </div>
</div>
`;
